### PR TITLE
Tide Implementation

### PR DESCRIPTION
**Pull request by alexcolpitts96**
Related to #1726 

Looking for a sanity check before documenting/testing.

### CODE REVIEW

**Comment by dennisbader** [comment]
Hi @alexcolpitts96 and thanks for this PR 🚀 
Yes, TiDE looks like a nice middle ground between the linear and transformer models.
Also nice that it supports all types of covariates and probabilistic forecasting.

If you want you can go ahead with the implementation :)

**Comment by dennisbader** [comment]
Thanks @alexcolpitts96, I'll review next week!

File: darts/tests/models/forecasting/test_tide.py
Comment (dennisbader) [inline darts/tests/models/forecasting/test_tide.py:41]: with current implementation this will not raise an error

**Review by dennisbader — CHANGES_REQUESTED** [review]
This is a really nice PR, thanks a lot @alexcolpitts96! 🚀 

At the current state I only have minor suggestions and:
- could you improve the TiDE tests? I see that currently it is basically a copy of the N/DLinear model tests which are not all applicable to the TiDE model.

For the next steps it would be ideal to implement the following:
- probabilistic support similar to how it's done in other models. We really want to offer probabilistic support for all TorchForecastingModels
- currently the model only works with future covariates. Can we adapt the model to support the case without future_covariates?
- could we add an embedding for catoegorical static covariates, similar to how we do it in TFTModel? Most of the time, static covariates are categorical and it would be great to have this support here as well.
- could you maybe test the model performance against nlinear/dlinear (not a unittest) and share the results with us? Would be nice to see how it performs. This can be a simply plot of predictions from TiDE against N/DLinear on some time series like AirPassengersDataset

**Comment by alexcolpitts96** [comment]
@dennisbader I am happy to make these changes. 

I haven't setup proper testing yet gradle was breaking for me. I have yet to resolve the issue. It keeps hanging on the formatting portion (black).

File: darts/tests/datasets/test_dataset_loaders.py
Comment (alexcolpitts96) [reply darts/tests/datasets/test_dataset_loaders.py:1]: It was a formatting issue. Black was complaining about it. When trying to run the tests it would hang for hours and then complain about it. I will try to re-run and get the message.

**Comment by dennisbader** [comment]
Let me know when it's ready for a second round review

**Comment by dennisbader** [comment]
Btw some tipps for testing / linting / formatting (if you aren't aware of this yet, as you mentioned testing with gradle and the linting issues):


- create a fresh env and activate it
- install the dev requirements from darts repo root
  ```
  pip install -r requirements/dev-all.txt
  ```
- install pre-commit hooks (with this, everything you commit something, it will check (and adapt) the code formatting and others)
  ```
  pre-commit install
  ```
- install darts locally
  ```
  pip install -e .
  ```
- run unit tests
  ```
  pytest darts/tests --no-cov
  ```

**Comment by alexcolpitts96** [comment]
@dennisbader I have added tests and probabilistic forecasts.

In general TiDE is an _OK_ model, in my opinion. It is more accurate than N/DLinear; however, it requires much more care in terms of training. Without gradient clipping, scheduled learning rates, and good initial learning rates it easily gets stuck in local minima. I also found that it is slower than N/DLinear and may not provide enough of an accuracy improvement to be relevant for short deadline forecasts.

**Comment by alexcolpitts96** [comment]
I COMPLETELY forgot that I was looking at dataset cleanup (#1539) before I started looking at TiDE.

I will try to strip out those commits.

**Review by alexcolpitts96 — COMMENTED** [review]
Pathing change

File: darts/models/forecasting/tide.py
Comment (dennisbader) [reply darts/models/forecasting/tide.py]: this has not been resolved yet

File: darts/models/forecasting/tide.py
Comment (dennisbader) [reply darts/models/forecasting/tide.py]: this has not been resolved yet

File: darts/models/forecasting/tide.py
Comment (dennisbader) [reply darts/models/forecasting/tide.py]: this has not been resolved yet

File: darts/models/forecasting/tide_model.py
Comment (dennisbader) [inline darts/models/forecasting/tide_model.py:280]: it would be nice to use similar parameter names as in the hyperparameter appendix from the paper.

**Review by dennisbader — CHANGES_REQUESTED** [review]
Great improvement @alexcolpitts96, thanks for that.

I had a couple of suggestions and a bug when using past covariates and no future covariates.

What I find a bit worrying is that the probabilistic predictions seem much better than the deterministic. In the top image below you see the default deterministic predictions and in the lower one the predicitons using QuantileRegression (the quantile intervals also seem a bit unstable).

From my comments from last time, only the categorical static covariates support is missing.

The unit tests should probably also be extended a bit. I would suggest to lean more towards the TFTModel tests rather than the D/NLinear tests.

Regarding the default hyperparmeters, the model is pretty slow (even slower than NBEATSModel, which shouldn't be the case in my opinion).
For example in the paper, they state that they only used `temporal_width=4` in contrast to our `temporal_width=128`

image

image

File: darts/models/forecasting/tide_model.py
Comment (Loudegaste) [inline darts/models/forecasting/tide_model.py:180]:
```suggestion
        self.decoders = nn.Sequential(
            *[
                _ResidualBlock(
                    input_dim=encoder_dim,
                    output_dim=encoder_dim,
                    hidden_size=hidden_size,
                    dropout=dropout,
                )
                for _ in range(num_decoder_layers)
            ],
            # add decoder output layer
            nn.Linear(
                encoder_dim,
                decoder_output_dim * self.output_chunk_length * self.nr_params,
            ),
```

File: darts/models/forecasting/tide_model.py
Comment (alexcolpitts96) [reply darts/models/forecasting/tide_model.py]: I like this change. I also 100% agree that the paper is very cryptic. They even have an error in the architecture diagram, since you can't stack the covariate projection with the decoder output.

File: darts/models/forecasting/tide.py
Comment (alexcolpitts96) [reply darts/models/forecasting/tide.py]: Changed it to temporal_decoder_hidden (this is the snake case equivalent of the paper's name).

File: darts/models/forecasting/tide_model.py
Comment (alexcolpitts96) [reply darts/models/forecasting/tide_model.py]: I had added this independently. I did it as a one liner, this is much more readable.

**Comment by alexcolpitts96** [comment]
@dennisbader I believe this is ready for another review. Sorry for this taking so long. 🙃 

- Added more tests. Focused on tests similar to TFT as you suggested.
- Added static covariate support
- Reduced default parameters to make model faster. This really bothers me with the model, it is more accurate; however, the compute cost incurred is pretty steep.

The biggest takeaway from your last review is that training was not always converging to good solutions. I found this model to converge to local minima without a ton of trainer parameters being set (gradient clipping, learning rate scheduling, etc). It ends up being more accurate, but it isn't as easy as NHiTS/NBEATS where you can throw anything into it and still get a reasonable solution out the other side.

I think I addressed all comments, but please let me know if I missed anything.

**Comment by dennisbader** [comment]
Hi @alexcolpitts96. Thanks again for your great job on the implementation of the TiDE model. I think you did an excellent job, it's a very clean PR, all comments were addressed, and the model is coherent with the paper. The paper also looks promising.

From the experiments we conducted, I have some concerns including the model to Darts, as it seems to be quite tricky to obtain good results. Until some convincing experiments can be done that show the strength/usefulness of this model over the others, I think it's better to wait. In the end, the quality of the end result and user experience should always prevail in case of doubt. 

What's your opinion on this? From your test cases with the trainer parameters, could you find a set of parameters that yielded robust results for different forecasting problems?

**Comment by alexcolpitts96** [comment]
@dennisbader I also have mixed feelings about TiDE. In my testing it has been better than any other model but has a performance penalty over N/DLinear and N-HiTS. I just changed companies this week so I don't have access to the experiments I ran, but I was able to get 8-12% MSE/MAE (I think?) better than N/DLinear and N-HiTS. Training was roughly 2-3x longer but inference wasn't nearly as computationally expensive. 

Training is more sensitive than other models but I found that as long as I used learning rate decay and used gradient clipping the results were always better. This does increase the burden on end users which I agree goes against the philosophy of darts as a package. The thing that I feel makes it a strong contender for being included in darts is that the model does handle anomalous samples really nicely (see figure from the paper). This would allow for end users to worry about anomalies as much in their inference data. 

image

If this doesn't get merged I have two things I will look into either:
1. Creating a package (darts-extended?) that has more advanced models that don't agree with the core use-case of darts.
2. Improving darts' custom model implementation (load in a torch model with some I/O schema and it does the rest?) and starting a repo containing the model definitions.

In the event that it doesn't get merged, I am more likely to keep pushing the implementation of Reversible Instance Normalization (#1865) and TSMixer (#1807) since they are both efficient and precise. RIN is specifically of interest since we can wrap the forward method of any of the models and get nearly "free" performance improvements (see the results in the issue). TSMixer is roughly an extension of N/DLinear with better performance.

TL;DR: I am ok with it either being merged or not.

**Comment by alexcolpitts96** [comment]
@dennisbader I have spent some more time thinking about this and feel slightly stronger about TiDE getting included in darts. It is distinctly different from other models in darts and would actually put darts ahead of Google in releasing TiDE in an easy to use package (it is supposed to be in [Vertex AI](https://cloud.google.com/vertex-ai/docs/tabular-data/forecasting/train-model), but I haven't been able to see it yet).

TiDE may be harder to train/tune than other models, but is unique in that it is the only MLP encoder-decoder model. I don't know if it was intentional, but there are generally 2-3 similar models in darts (Transformer and TFT, DLinear and NLinear, NHiTS and NBEATS, etc). I am now thinking that having TiDE and TSMixer (#1807) would give darts strong offerings in attention-free MLP encoder-decoders. These models would also be a good start for looking at using reversible instance normalization.

**Comment by dennisbader** [comment]
Hey @alexcolpitts96 and thanks for the update. That sounds great, I'm glad that you found a way to improve the model performance :) 🚀  In that case I think we can go for including this in the next Darts release. I can do another round of review later today/tomorrow.

One thing that would be nice now that it'll be integrated, is to add a _simple_ example notebook for the model similar to other TorchForecastingModels (darts/examples). Do you think you could add one?

**Comment by alexcolpitts96** [comment]
@dennisbader adding an example shouldn't be too tricky. I essentially built one already for testing it.

I would hold off on doing another review until after the RINorm PR is merged and I add it to TiDE. 

At that point it should have all the features that the original paper has (+probabalistic forecasting) and would be ready for another review.

**Comment by dennisbader** [comment]
@alexcolpitts96, perfect 👍  I just merged the RINorm. Let me know when this one is ready for another review 🚀

File: darts/models/forecasting/tide_model.py
Comment (dennisbader) [inline darts/models/forecasting/tide_model.py:334]: we can avoid iterating over nr_params if we expand the dimension of `stdev`, `mean`, `affine_weight`, `affine_bias` in `RINorm.inverse` (all models should have output shape `(batch_size, output_chunk_length, n_targets, nr_params)`).

just as an example:
```
x = x * self.stdev.view(self.stdev.shape + (1,))
```

then we can call `inverse` directly:
```suggestion
            y = self.rin.inverse(y)
```

**Review by dennisbader — CHANGES_REQUESTED** [review]
Very nice @alexcolpitts96 👍  I also played around a bit more with the model today and it feels more robust now 🚀 
I had minor suggestions and a question about covariates normalization that I'd like to know your opinion on.

From what I see, the only things remaining are:
- TiDE example notebook
- categorical static covariates through embeddings (this one we can also open another PR for, as we can do it at the same time for N/DLinear as well).

File: darts/models/forecasting/tide_model.py
Comment (alexcolpitts96) [reply darts/models/forecasting/tide_model.py:334]: I can adjust the input to just consider the target variables. How to handle covariates is something that I had spent some time thinking about as well.

I think we should stick with just the target. It will be easy to shoot yourself in the foot if we enforce that all covariates be adjusted.

**Comment by alexcolpitts96** [comment]
@dennisbader assuming the tests pass, this should be good to go🤞

**Review by dennisbader — APPROVED** [review]
Congratulations @alexcolpitts96 on this great PR, it will be a very cool addition to darts :) 🚀 💯 

The effort that went into this is really appreciated, thanks a lot!

I updated some last things to make it ready for the merge:
- updated the example notebook (just some slight rephrasing to match the style of other Darts documents).
- Added TiDE model to the model sections of the documentation
- some additions to the documentation

Let me know if you want to have a look at it :) If not, then it's ready to be merged 🚀

**Comment by alexcolpitts96** [comment]
I don't have any issues with it. Thanks for all the help getting it cleaned up and improved. 😁

**Pull request merged**
